### PR TITLE
JNI: Add generateListOffsets API

### PR DESCRIPTION
**Pull request by sperlingxx**
Add generateListOffsets API, converting list lengths to list offsets, which is useful in the development of spark-rapids.

For example, the support of [array_repeat](https://github.com/NVIDIA/spark-rapids/issues/5226) and [arrays_zip](https://github.com/NVIDIA/spark-rapids/issues/5229) relies on this API.

### CODE REVIEW

File: java/src/main/native/src/ColumnViewJni.cu [hunk 1]
```diff
@@ -15,8 +15,11 @@
  */
 
 #include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
 #include <cudf/detail/iterator.cuh>
 #include <cudf/detail/valid_if.cuh>
+#include <rmm/exec_policy.hpp>
+#include <thrust/scan.h>
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cu:22]:
```suggestion

#include <rmm/exec_policy.hpp>

#include <thrust/scan.h>
```

File: java/src/main/native/src/ColumnViewJni.cu [hunk 2]
```diff
@@ -51,4 +54,18 @@ new_column_with_boolean_column_as_validity(cudf::column_view const &exemplar,
...
 }
 
+std::unique_ptr<cudf::column> generate_list_offsets(cudf::column_view const &list_length,
+                                                    rmm::cuda_stream_view stream) {
+  CUDF_EXPECTS(list_length.type().id() == cudf::type_id::INT32,
+               "Input column does not have type INT32.");
+
+  auto offsets = cudf::make_numeric_column(cudf::data_type{cudf::type_id::INT32},
+                                           list_length.size() + 1, cudf::mask_state::UNALLOCATED);
+  auto output_view = offsets->mutable_view();
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cu:64]:
```suggestion
```

Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cu:64]:
```suggestion
  auto const output_view = offsets->mutable_view();
```

File: java/src/main/native/src/ColumnViewJni.cu [hunk 3]
```diff
@@ -51,4 +54,18 @@ new_column_with_boolean_column_as_validity(cudf::column_view const &exemplar,
...
+  thrust::inclusive_scan(rmm::exec_policy(stream), list_length.template begin<offset_type>(),
+                         list_length.template end<offset_type>(),
+                         output_view.template begin<offset_type>() + 1);
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cu:67]: This will cause undefined behavior, since it will leave the first offset value uninitialized.

Comment (ttnghia) [reply java/src/main/native/src/ColumnViewJni.cu:67]: In addition, the entire function `generate_lists_offsets` has been implemented in cudf. It is in the `include/cudf/string/detail/utilities.cuh`. We can try to expose it to be accessed here.

Comment (ttnghia) [reply java/src/main/native/src/ColumnViewJni.cu:67]: See: https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/include/cudf/strings/detail/utilities.cuh#L52

File: java/src/main/native/src/ColumnViewJni.cu (hunk 1 above)
Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu:23]: The clang format script deleted these two blank lines = =

File: java/src/main/native/src/ColumnViewJni.cu (hunk 3 above)
Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu]: I replaced this implementation with a call of `make_offsets_child_column`.

**Review by ttnghia — APPROVED** [review]

**Comment by sperlingxx** [comment]
@gpucibot merge

File: java/src/main/native/src/ColumnViewJni.cu (hunk 1 above)
Comment (ttnghia) [reply java/src/main/native/src/ColumnViewJni.cu:23]: Wait, since we removed their usage, let's remove them too.

Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu]: Yeah, I am so careless T_T

Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu]: Removed them

**Review by ttnghia — APPROVED** [review]

File: java/src/main/native/src/ColumnViewJni.cu [hunk 4]
```diff
@@ -15,8 +15,10 @@
  */
 
 #include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
 #include <cudf/detail/iterator.cuh>
 #include <cudf/detail/valid_if.cuh>
+#include <cudf/strings/detail/utilities.cuh>
```
Comment (jrhemstad) [inline java/src/main/native/src/ColumnViewJni.cu:21]: If you're having to include a header from a `detail` folder, something is wrong. These can break at any time, for any reason, and there will be no sympathy for any code that breaks.

File: java/src/main/native/src/ColumnViewJni.cu [hunk 5]
```diff
@@ -51,4 +53,14 @@ new_column_with_boolean_column_as_validity(cudf::column_view const &exemplar,
...
 
+std::unique_ptr<cudf::column> generate_list_offsets(cudf::column_view const &list_length,
+                                                    rmm::cuda_stream_view stream) {
+  CUDF_EXPECTS(list_length.type().id() == cudf::type_id::INT32,
+               "Input column does not have type INT32.");
+
+  auto const begin_iter = list_length.template begin<cudf::size_type>();
+  auto const end_iter = list_length.template end<cudf::size_type>();
+
+  return cudf::strings::detail::make_offsets_child_column(begin_iter, end_iter, stream);
```
Comment (jrhemstad) [inline java/src/main/native/src/ColumnViewJni.cu:64]: There's no reason to use a detail API here. This is just a scan and can use `cudf::scan`.

Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu:64]: Hi @jrhemstad, we need to allocate a output buffer whose size is N + 1. Then, project the result of inclusive scan to the output buffer (with the offset 1).
Is `thrust::inclusive_scan` more flexible to do this job than `cudf::scan`?

Comment (sperlingxx) [reply java/src/main/native/src/ColumnViewJni.cu:64]: If I am not wrong, we need to run an extra memcpy with the `cudf::scan`.

Comment (ttnghia) [reply java/src/main/native/src/ColumnViewJni.cu:64]: Humm, if we can't use the detail API then I hate to recommend this: just copy exactly the code from `cudf::strings::detail::make_offsets_child_column`.

File: java/src/main/native/src/ColumnViewJni.cu [hunk 6]
```diff
@@ -15,9 +15,12 @@
  */
 
 #include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
 #include <cudf/detail/iterator.cuh>
 #include <cudf/detail/valid_if.cuh>
 
+#include "rmm/exec_policy.hpp"
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cu:22]:
```suggestion
#include <rmm/exec_policy.hpp>
```

File: java/src/main/java/ai/rapids/cudf/ColumnView.java [hunk 7]
```diff
@@ -3473,6 +3473,16 @@ public final ColumnVector listSortRows(boolean isDescending, boolean isNullSmall
     return new ColumnVector(listSortRows(getNativeView(), isDescending, isNullSmallest));
   }
 
+  /**
+   * Generate list offsets from sizes of each list.
+   * NOTICE: This column should type in INT32. And no null and negative value is allowed.
```
Comment (firestarman) [inline java/src/main/java/ai/rapids/cudf/ColumnView.java:3478]: Is this requirement for the input column ? What if it is not matched ? Better to add comment for it.

Comment (sperlingxx) [reply java/src/main/java/ai/rapids/cudf/ColumnView.java]: Fixed

**Review by firestarman — APPROVED** [review]

**Comment by sperlingxx** [comment]
rerun tests

**Comment by sperlingxx** [comment]
rerun tests

**Comment by sperlingxx** [comment]
@gpucibot merge

**Pull request merged**